website/blog/metric_labels.md

```diff
@@ -0,0 +1,130 @@
+`path` and `status_code` labels using [Prometheus' Go API][prometheus_api]:
+
+```go
+requests := prometheus.NewCounterVec(
+    prometheus.CounterOpts{
+        Name: "http_requests",
+        Help: "Number of HTTP requests.",
+    },
+    []string{"path", "status_code"},
+)
+```
+
+And here is how to increment the counter:
+
+```go
+requests.With(prometheus.Labels{"path": "/foo", "status_code": "404"}).Inc()
+```
+
+Similarly, here is how to declare the same metric using [OpenTelemetry's Go
+API][otel_api]:
+
+```go
+requests, err := meter.Int64Counter(
+    "http_requests",
+    metric.WithDescription("Number of HTTP requests."),
+)
+```
+
+Note that there isn't a way for us to specify the `path` and `status_code`
+labels when we declare the counter. Instead, we provide the labels (what
+OpenTelemtry calls attributes) when we increment the counter:
+
+```go
+requests.Add(ctx, 1, metric.WithAttributes(
+    attribute.Key("path").String("/foo"),
+    attribute.Key("status_code").Int(404),
+))
+```
+
+## Drawbacks
+
+Specifying labels as a set of key-value pairs like this has a number of
+disadvantages.
+
+- **Misspelling.** You can misspell label names. For example, you might
+  accidentally use label name `"stats_code"` instead of `"status_code"`.
+- **Mistyping.** You can use label values with the wrong type. For example, you
+  might accidentally pass a status like `"404 Not Found"` instead of the status
+  code `404`.
+- **Misremembering.** You can misremember the name of a label completely. Was it
+  `"path"` or `"endpoint"` or `"URL"`? You can accidentally use the wrong name.
+- **Too many labels.** You can pass a label that wasn't defined on the metric.
+  If we remove the `status_code` label, for example, we have to be careful to
+  update our code everywhere that uses it.
+- **Too few labels.** You can forget to pass a label. If we add a new label to a
+  metric, for example, we again have to carefully update our code to use it.
+
+## Service Weaver's Approach
+
+Service Weaver avoids these drawbacks by integrating metric labels with Go's
+type system. Specifically, Service Weaver represents metric labels as structs
+and uses generics to instantiate metrics. Here's how to declare the
+`http_requests` metric using [Service Weaver's API][weaver_api]:
+
+```go
+type labels struct {
+    Path       string `weaver:"path"`
+    StatusCode int    `weaver:"status_code"`
+}
+
+var requests = metrics.NewCounterMap[labels](
+    "http_requests",
+    "Number of HTTP requests.",
+)
+```
+
+And here's how to increment the counter:
+
+```go
+requests.Get(labels{Path: "/foo", StatusCode: 404}).Inc()
+```
+
+Labels can be any struct where every field is a string, bool, or integer.
+`NewCounterMap` panics if you call it with an invalid label struct. By
+leveraging Go's type system, Service Weaver's API avoids the disadvantages of
+key-value labels described above.
+
+- **Misspelling.** If you misspell a field name, your code will not compile
```

Comment:
Maybe the code looks more complex as another disadvantage? Also, it doesn't age well? although this might refer to your too few/too many labels

Reply:
Great point. I agree our label based approach is a bit simpler, but since that's a bit subjective, I figure it's best to stick to some of the more objective advantages/disadvantages?